File: jam-safrole/src/main/kotlin/io/forge/jam/safrole/accumulation/AccumulationTypes.kt
package io.forge.jam.safrole.accumulation

import io.forge.jam.core.ExecutionResult
import io.forge.jam.core.JamByteArray
import io.forge.jam.core.WorkReport
import io.forge.jam.safrole.report.ServiceInfo

/**
 * Operand tuple as defined in Gray Paper equation 106.
 * Contains extracted work item data combined with work report context.
 */
data class OperandTuple(
    val packageHash: JamByteArray,     // Work package hash
    val segmentRoot: JamByteArray,     // Segment root from availability spec
    val authorizerHash: JamByteArray,  // Authorizer hash
    val payloadHash: JamByteArray,     // Work item payload hash
    val gasLimit: Long,                // Gas limit for accumulation
    val authTrace: JamByteArray,       // Authorizer trace output
    val result: ExecutionResult        // Refinement result (blob or error)
)

/**
 * Deferred transfer as defined in Gray Paper equation 117.
 * Represents a transfer queued during accumulation for processing in next iteration.
 */
data class DeferredTransfer(
    val source: Long,       // Source service index
    val destination: Long,  // Destination service index
    val amount: Long,       // Balance to transfer
    val memo: JamByteArray, // Memo (128 bytes)
    val gasLimit: Long      // Gas limit for on_transfer handler
) {
    companion object {
        const val MEMO_SIZE = 128
    }
}

/**
 * Accumulation input - union of OperandTuple or DeferredTransfer
 * as defined in Gray Paper equation 126.
 */
sealed class AccumulationOperand {
    abstract fun encode(): ByteArray

    data class WorkItem(val operand: OperandTuple) : AccumulationOperand() {
        override fun encode(): ByteArray {
            val op = operand

            // AccumulationInput::OperandTuple (Variant 0)
            // Standard JAM enum encoding (1 byte).
            val variant = byteArrayOf(0)

            // GasLimit - JAM compact encoded
            val gasLimitBytes = io.forge.jam.core.encodeCompactInteger(op.gasLimit)

            // WorkResult - variant 0 + JAM-compact(len) + bytes for success, or variant for error
            val ok = op.result.ok
            val resultBytes = if (ok != null) {
                val len = io.forge.jam.core.encodeCompactInteger(ok.bytes.size.toLong())
                byteArrayOf(0) + len + ok.bytes // Tag 0 for Success
            } else {
                byteArrayOf(2) // Tag 2 for Panic (or other error)
            }

            // AuthTrace (JAM-compact(len) + bytes)
            val authTraceLen = io.forge.jam.core.encodeCompactInteger(op.authTrace.bytes.size.toLong())

            // Order: package, segRoot, authorizer, payload, gasLimit, workResult, authorizerTrace
            return variant + op.packageHash.bytes + op.segmentRoot.bytes + op.authorizerHash.bytes +
                op.payloadHash.bytes + gasLimitBytes + resultBytes + authTraceLen + op.authTrace.bytes
        }
    }

    data class Transfer(val transfer: DeferredTransfer) : AccumulationOperand() {
        override fun encode(): ByteArray {
            // Prepend tag 1 for Transfer
            // sender/destination are ServiceIndex (UInt32), amount/gasLimit are Balance/Gas (UInt64)
            return byteArrayOf(1) +
                encodeU32(transfer.source.toInt()) +
                encodeU32(transfer.destination.toInt()) +
                encodeLong(transfer.amount) +
                transfer.memo.bytes +
                encodeLong(transfer.gasLimit)
        }

        private fun encodeU32(value: Int): ByteArray {
            return ByteArray(4) { i -> ((value shr (i * 8)) and 0xFF).toByte() }
        }
    }

    companion object {
        fun encodeLong(value: Long): ByteArray {
            return ByteArray(8) { i -> ((value shr (i * 8)) and 0xFF).toByte() }
        }
    }
}

/**
 * Partial state as defined in Gray Paper equation 133.
 * Contains state components both needed and mutable by the accumulation process.
 */
data class PartialState(
    val accounts: MutableMap<Long, ServiceAccount>,
    val stagingSet: MutableList<JamByteArray>,           // Validator keys
    val authQueue: MutableList<MutableList<JamByteArray>>, // Per-core auth queues
    var manager: Long,                                    // Manager service ID
    val assigners: MutableList<Long>,                    // Per-core assigners
    var delegator: Long,                                 // Delegator service ID
    var registrar: Long,                                 // Registrar service ID
    val alwaysAccers: MutableMap<Long, Long>             // Always-accumulate services -> gas
) {
    fun deepCopy(): PartialState {
        return PartialState(
            accounts = accounts.mapValues { it.value.copy() }.toMutableMap(),
            stagingSet = stagingSet.map { it.copy() }.toMutableList(),
            authQueue = authQueue.map { it.map { h -> h.copy() }.toMutableList() }.toMutableList(),
            manager = manager,
            assigners = assigners.toMutableList(),
            delegator = delegator,
            registrar = registrar,
            alwaysAccers = alwaysAccers.toMutableMap()
        )
    }
}

/**
 * Service account combining service info with mutable storage and preimages.
 */
data class ServiceAccount(
    val info: ServiceInfo,
    val storage: MutableMap<JamByteArray, JamByteArray>,    // Key -> Value storage
    val preimages: MutableMap<JamByteArray, JamByteArray>,  // Hash -> Blob preimages
    val preimageRequests: MutableMap<PreimageKey, PreimageRequest>, // Requested preimages
    var lastAccumulated: Long = 0                            // Last accumulation timestamp
) {
    fun copy(): ServiceAccount {
        return ServiceAccount(
            info = info.copy(),
            storage = storage.toMutableMap(),
            preimages = preimages.toMutableMap(),
            preimageRequests = preimageRequests.toMutableMap(),
            lastAccumulated = lastAccumulated
        )
    }
}

/**
 * Key for preimage requests (hash + length).
 */
data class PreimageKey(
    val hash: JamByteArray,
    val length: Int
)

/**
 * Preimage request state.
 */
data class PreimageRequest(
    val requestedAt: List<Long> // Timestamps when requested
)

/**
 * Result of single-service accumulation as defined in Gray Paper equation 291.
 */
data class AccumulationOneResult(
    val postState: PartialState,                          // Modified state
    val deferredTransfers: List<DeferredTransfer>,        // Outgoing transfers
    val yield: JamByteArray?,                             // Accumulation output (32-byte hash or null)
    val gasUsed: Long,                                    // Actual gas consumed
    val provisions: Set<Pair<Long, JamByteArray>>         // Service/blob pairs to provision
)

/**
 * Result of parallel accumulation.
 */
data class AccumulationParResult(
    val postState: PartialState,
    val deferredTransfers: List<DeferredTransfer>,
    val outputs: Map<Long, JamByteArray>,                 // Service -> output hash
    val gasUsed: List<Pair<Long, Long>>                   // Service -> gas used
)

/**
 * Result of sequential accumulation.
 */
data class AccumulationSeqResult(
    val reportsAccumulated: Int,
    val postState: PartialState,
    val outputs: Map<Long, JamByteArray>,
    val gasUsed: List<Pair<Long, Long>>
)

/**
 * Execution exit reason for PVM.
 */
enum class ExitReason {
    HALT,        // Normal completion
    PANIC,       // Panic (use checkpoint state)
    OUT_OF_GAS,  // Gas exhausted
    PAGE_FAULT,  // Memory access error
    HOST_CALL,   // Awaiting host call response
    INVALID_CODE // Code compilation failed
}

/**
 * Accumulation context managing dual state (x for normal, y for checkpoint).
 */
class AccumulationContext(
    var x: PartialState,  // Normal execution state
    var y: PartialState,  // Checkpoint state (used on panic)
    val serviceIndex: Long,
    val timeslot: Long,
    val entropy: JamByteArray,
    val deferredTransfers: MutableList<DeferredTransfer> = mutableListOf(),
    val provisions: MutableSet<Pair<Long, JamByteArray>> = mutableSetOf()
) {
    /**
     * Checkpoint: copy current state x to checkpoint y.
     */
    fun checkpoint() {
        y = x.deepCopy()
    }

    /**
     * Collapse: select final state based on exit reason.
     * On panic, revert to checkpoint state y.
     */
    fun collapse(exitReason: ExitReason): PartialState {
        return when (exitReason) {
            ExitReason.PANIC -> y
            ExitReason.INVALID_CODE -> {
                x.accounts.remove(serviceIndex)
                x
            }
            else -> x
        }
    }
}

/**
 * Extract operand tuples from work reports for a specific service.
 */
fun extractOperandTuples(reports: List<WorkReport>, serviceId: Long): List<OperandTuple> {
    return reports.flatMap { report ->
        report.results
            .filter { it.serviceId == serviceId }
            .map { result ->
                OperandTuple(
                    packageHash = report.packageSpec.hash,
                    segmentRoot = report.packageSpec.erasureRoot,
                    authorizerHash = report.authorizerHash,
                    payloadHash = result.payloadHash,
                    gasLimit = result.accumulateGas,
                    authTrace = report.authOutput,
                    result = result.result
                )
            }
    }
}

/**
 * Calculate total gas limit for a service from work reports and transfers.
 */
fun calculateServiceGasLimit(
    reports: List<WorkReport>,
    transfers: List<DeferredTransfer>,
    freeGas: Map<Long, Long>,
    serviceId: Long
): Long {
    val freeGasAmount = freeGas[serviceId] ?: 0L
    val transferGas = transfers.filter { it.destination == serviceId }.sumOf { it.gasLimit }
    val workItemGas = reports.flatMap { r -> r.results.filter { it.serviceId == serviceId } }
        .sumOf { it.accumulateGas }
    return freeGasAmount + transferGas + workItemGas
}

/**
 * E function (Gray Paper eq. 49-61): Edit ready queue by removing accumulated reports
 * and pruning fulfilled dependencies from remaining records.
 */
fun editReadyQueue(
    queue: List<ReadyRecord>,
    accumulatedHashes: Set<JamByteArray>
): List<ReadyRecord> {
    return queue
        // Remove records whose report has already been accumulated
        .filter { record -> record.report.packageSpec.hash !in accumulatedHashes }
        // Remove fulfilled dependencies from remaining records
        .map { record ->
            ReadyRecord(
                report = record.report,
                dependencies = record.dependencies.filter { it !in accumulatedHashes }
            )
        }
}

/**
 * Q function (Gray Paper eq. 63-73): Extract accumulatable reports via topological sort.
 * Returns reports that can be accumulated (all dependencies satisfied) and remaining queued records.
 */
fun extractAccumulatableReports(
    queue: List<ReadyRecord>,
    initiallyAccumulated: Set<JamByteArray>
): Pair<List<WorkReport>, List<ReadyRecord>> {
    val accumulated = initiallyAccumulated.toMutableSet()
    val result = mutableListOf<WorkReport>()
    var remaining = queue.toList()

    do {
        val (ready, notReady) = remaining.partition { record ->
            record.dependencies.all { it in accumulated }
        }
        if (ready.isEmpty()) break

        ready.forEach { record ->
            result.add(record.report)
            accumulated.add(record.report.packageSpec.hash)
        }
        remaining = notReady
    } while (true)

    return Pair(result, remaining)
}

/**
 * Count work items per service from accumulated reports.
 */
fun countWorkItemsPerService(reports: List<WorkReport>): Map<Long, Int> {
    return reports.flatMap { report ->
        report.results.map { it.serviceId }
    }.groupingBy { it }.eachCount()
}

File: modules/protocol/src/main/scala/io/forge/jam/protocol/accumulation/AccumulationTypes.scala
package io.forge.jam.protocol.accumulation

import io.forge.jam.core.JamBytes
import io.forge.jam.core.scodec.JamCodecs
import io.forge.jam.core.primitives.Hash
import io.forge.jam.core.types.service.ServiceInfo
import io.forge.jam.core.types.work.ExecutionResult
import spire.math.{UInt, ULong}

import scala.collection.mutable

/**
 * Contains extracted work item data combined with work report context.
 *
 * @param packageHash Work package hash (32 bytes)
 * @param segmentRoot Segment root from availability spec (32 bytes)
 * @param authorizerHash Authorizer hash (32 bytes)
 * @param payloadHash Work item payload hash (32 bytes)
 * @param gasLimit Gas limit for accumulation
 * @param authTrace Authorizer trace output (variable length)
 * @param result Refinement result (blob or error)
 */
final case class OperandTuple(
  packageHash: JamBytes,
  segmentRoot: JamBytes,
  authorizerHash: JamBytes,
  payloadHash: JamBytes,
  gasLimit: Long,
  authTrace: JamBytes,
  result: ExecutionResult
)

/**
 * Deferred transfer
 * Represents a transfer queued during accumulation for processing in next iteration.
 *
 * @param source Source service index
 * @param destination Destination service index
 * @param amount Balance to transfer
 * @param memo Memo (128 bytes fixed size)
 * @param gasLimit Gas limit for on_transfer handler
 */
final case class DeferredTransfer(
  source: Long,
  destination: Long,
  amount: Long,
  memo: JamBytes,
  gasLimit: Long
)

object DeferredTransfer:
  val MEMO_SIZE: Int = 128

/**
 * Accumulation input - union of OperandTuple or DeferredTransfer
 */
sealed trait AccumulationOperand:
  /**
   * Encode the operand to bytes using Gray Paper natural encoding.
   */
  def encode(): Array[Byte]

object AccumulationOperand:
  final case class WorkItem(operand: OperandTuple) extends AccumulationOperand:
    override def encode(): Array[Byte] =
      val op = operand

      val variant = JamCodecs.encodeCompactInteger(0)
      val gasLimitBytes = JamCodecs.encodeCompactInteger(op.gasLimit)
      val resultBytes = op.result match
        case ExecutionResult.Ok(output) =>
          val len = JamCodecs.encodeCompactInteger(output.length.toLong)
          Array[Byte](0) ++ len ++ output.toArray // Tag 0 for Success (UInt8)
        case ExecutionResult.OOG => Array[Byte](1)
        case ExecutionResult.Panic => Array[Byte](2)
        case ExecutionResult.BadExports => Array[Byte](3)
        case ExecutionResult.Oversize => Array[Byte](4)
        case ExecutionResult.BadCode => Array[Byte](5)
        case ExecutionResult.CodeTooLarge => Array[Byte](6)

      val authTraceLen = JamCodecs.encodeCompactInteger(op.authTrace.length.toLong)
      variant ++
        op.packageHash.toArray ++ op.segmentRoot.toArray ++ op.authorizerHash.toArray ++
        op.payloadHash.toArray ++ gasLimitBytes ++ resultBytes ++ authTraceLen ++ op.authTrace.toArray

  /**
   * Transfer operand containing a DeferredTransfer.
   * Encoded as variant 1.
   */
  final case class Transfer(transfer: DeferredTransfer) extends AccumulationOperand:
    override def encode(): Array[Byte] =
      JamCodecs.encodeCompactInteger(1) ++
        JamCodecs.encodeU32LE(UInt(transfer.source.toInt)) ++
        JamCodecs.encodeU32LE(UInt(transfer.destination.toInt)) ++
        JamCodecs.encodeU64LE(ULong(transfer.amount)) ++
        transfer.memo.toArray ++
        JamCodecs.encodeU64LE(ULong(transfer.gasLimit))

/**
 * Key for preimage requests (hash + length).
 *
 * @param hash Hash of the preimage (32 bytes)
 * @param length Expected length of the preimage
 */
final case class PreimageKey(
  hash: Hash,
  length: Int
)

/**
 * Preimage request state.
 *
 * @param requestedAt Timestamps (timeslots) when requested (0-3 entries)
 */
final case class PreimageRequest(
  requestedAt: List[Long]
)

/**
 * Service account combining service info with mutable storage and preimages.
 * Used during accumulation to track state changes.
 *
 * @param info Service info containing balance, code hash, gas limits, etc.
 * @param storage Key-value storage (mutable)
 * @param preimages Hash to blob mapping (mutable)
 * @param preimageRequests Requested preimages (mutable)
 * @param lastAccumulated Last accumulation timestamp
 */
final case class ServiceAccount(
  info: ServiceInfo,
  storage: mutable.Map[JamBytes, JamBytes],
  preimages: mutable.Map[Hash, JamBytes],
  preimageRequests: mutable.Map[PreimageKey, PreimageRequest],
  var lastAccumulated: Long = 0
):
  /**
   * Create a deep copy of this service account.
   * All mutable collections are copied to ensure independence.
   */
  def copy(): ServiceAccount =
    ServiceAccount(
      info = info.copy(),
      storage = mutable.Map.from(storage),
      preimages = mutable.Map.from(preimages),
      preimageRequests = mutable.Map.from(preimageRequests),
      lastAccumulated = lastAccumulated
    )

object ServiceAccount:
  /**
   * Create an empty service account with default values.
   */
  def empty(info: ServiceInfo): ServiceAccount =
    ServiceAccount(
      info = info,
      storage = mutable.Map.empty,
      preimages = mutable.Map.empty,
      preimageRequests = mutable.Map.empty,
      lastAccumulated = 0
    )

/**
 * Mutable subset of JAM state used during accumulation.
 * Contains state components both needed and mutable by the accumulation process.
 *
 * @param accounts Service accounts by index
 * @param stagingSet Validator keys for staging
 * @param authQueue Per-core authorization queues
 * @param manager Manager service ID
 * @param assigners Per-core assigners
 * @param delegator Delegator service ID
 * @param registrar Registrar service ID
 * @param alwaysAccers Always-accumulate services to gas mapping
 * @param rawServiceDataByStateKey Raw state data lookups
 * @param rawServiceAccountsByStateKey Raw account lookups
 */
final case class PartialState(
  accounts: mutable.Map[Long, ServiceAccount],
  stagingSet: mutable.ListBuffer[JamBytes],
  authQueue: mutable.ListBuffer[mutable.ListBuffer[JamBytes]],
  var manager: Long,
  assigners: mutable.ListBuffer[Long],
  var delegator: Long,
  var registrar: Long,
  alwaysAccers: mutable.Map[Long, Long],
  rawServiceDataByStateKey: mutable.Map[JamBytes, JamBytes] = mutable.Map.empty,
  rawServiceAccountsByStateKey: mutable.Map[JamBytes, JamBytes] = mutable.Map.empty
):
  /**
   * Create a deep copy of this partial state.
   * All mutable collections and contained objects are copied to ensure independence.
   */
  def deepCopy(): PartialState =
    PartialState(
      accounts = mutable.Map.from(accounts.view.mapValues(_.copy())),
      stagingSet = mutable.ListBuffer.from(stagingSet.map(b => JamBytes(b.toArray))),
      authQueue = mutable.ListBuffer.from(authQueue.map(q => mutable.ListBuffer.from(q.map(h => JamBytes(h.toArray))))),
      manager = manager,
      assigners = mutable.ListBuffer.from(assigners),
      delegator = delegator,
      registrar = registrar,
      alwaysAccers = mutable.Map.from(alwaysAccers),
      rawServiceDataByStateKey = mutable.Map.from(rawServiceDataByStateKey),
      rawServiceAccountsByStateKey = mutable.Map.from(rawServiceAccountsByStateKey)
    )

object PartialState:
  /**
   * Create an empty partial state with default values.
   */
  def empty: PartialState =
    PartialState(
      accounts = mutable.Map.empty,
      stagingSet = mutable.ListBuffer.empty,
      authQueue = mutable.ListBuffer.empty,
      manager = 0L,
      assigners = mutable.ListBuffer.empty,
      delegator = 0L,
      registrar = 0L,
      alwaysAccers = mutable.Map.empty
    )

/**
 * Execution exit reason for PVM.
 * Determines whether to use normal state (x) or checkpoint state (y).
 */
enum ExitReason:
  /** Normal completion - use normal state x */
  case HALT

  /** Panic - use checkpoint state y */
  case PANIC

  /** Gas exhausted - use checkpoint state y */
  case OUT_OF_GAS

  /** Memory access error - use checkpoint state y */
  case PAGE_FAULT

  /** Awaiting host call response */
  case HOST_CALL

  /** Code compilation failed - use checkpoint state y */
  case INVALID_CODE

/**
 * Accumulation context managing dual state (x for normal, y for checkpoint).
 * Provides checkpoint and collapse operations for accumulation.
 *
 * @param x Normal execution state
 * @param y Checkpoint state (used on panic)
 * @param serviceIndex Current service being accumulated
 * @param timeslot Current timeslot
 * @param entropy Entropy for the epoch
 * @param deferredTransfers Transfers queued during accumulation (normal state)
 * @param deferredTransfersCheckpoint Checkpoint of deferred transfers
 * @param provisions Preimage provisions (normal state)
 * @param provisionsCheckpoint Checkpoint of provisions
 * @param yieldHash Accumulation output hash (normal state)
 * @param yieldCheckpoint Checkpoint of yield hash
 * @param nextAccountIndex Next available service account index
 * @param minPublicServiceIndex Minimum public service index (S_S from Gray Paper, 2^16)
 */
final class AccumulationContext(
  var x: PartialState,
  var y: PartialState,
  val serviceIndex: Long,
  val timeslot: Long,
  val entropy: JamBytes,
  val deferredTransfers: mutable.ListBuffer[DeferredTransfer] = mutable.ListBuffer.empty,
  val deferredTransfersCheckpoint: mutable.ListBuffer[DeferredTransfer] = mutable.ListBuffer.empty,
  val provisions: mutable.Set[(Long, JamBytes)] = mutable.Set.empty,
  val provisionsCheckpoint: mutable.Set[(Long, JamBytes)] = mutable.Set.empty,
  var yieldHash: Option[JamBytes] = None,
  var yieldCheckpoint: Option[JamBytes] = None,
  var nextAccountIndex: Long = 65536L,
  val minPublicServiceIndex: Long = 65536L
):

  /**
   * Checkpoint: copy current state x to checkpoint y, including yield, provisions, and transfers.
   */
  def checkpoint(): Unit =
    y = x.deepCopy()
    yieldCheckpoint = yieldHash
    provisionsCheckpoint.clear()
    provisionsCheckpoint ++= provisions
    deferredTransfersCheckpoint.clear()
    deferredTransfersCheckpoint ++= deferredTransfers

  /**
   * Collapse: select final state based on exit reason.
   * On panic, out of gas, page fault, or invalid code, revert to checkpoint state y.
   *
   * @param exitReason The reason for execution termination
   * @return The appropriate state (y for error conditions, x otherwise)
   */
  def collapse(exitReason: ExitReason): PartialState =
    exitReason match
      case ExitReason.PANIC | ExitReason.OUT_OF_GAS | ExitReason.PAGE_FAULT | ExitReason.INVALID_CODE => y
      case _ => x

  /**
   * Get provisions based on exit reason.
   * On panic or out of gas, use checkpoint provisions.
   *
   * @param exitReason The reason for execution termination
   * @return The appropriate provisions set
   */
  def getProvisions(exitReason: ExitReason): Set[(Long, JamBytes)] =
    exitReason match
      case ExitReason.PANIC | ExitReason.OUT_OF_GAS => provisionsCheckpoint.toSet
      case _ => provisions.toSet

  /**
   * Get deferred transfers based on exit reason.
   * On panic or out of gas, use checkpoint transfers.
   *
   * @param exitReason The reason for execution termination
   * @return The appropriate list of deferred transfers
   */
  def getDeferredTransfers(exitReason: ExitReason): List[DeferredTransfer] =
    exitReason match
      case ExitReason.PANIC | ExitReason.OUT_OF_GAS => deferredTransfersCheckpoint.toList
      case _ => deferredTransfers.toList

object AccumulationContext:
  /**
   * Create a new accumulation context with the given parameters.
   * Both x and y states are initialized with the same content (deep copied).
   */
  def apply(
    initialState: PartialState,
    serviceIndex: Long,
    timeslot: Long,
    entropy: JamBytes,
    nextAccountIndex: Long = 65536L,
    minPublicServiceIndex: Long = 65536L
  ): AccumulationContext =
    new AccumulationContext(
      x = initialState.deepCopy(),
      y = initialState.deepCopy(),
      serviceIndex = serviceIndex,
      timeslot = timeslot,
      entropy = entropy,
      nextAccountIndex = nextAccountIndex,
      minPublicServiceIndex = minPublicServiceIndex
    )

/**
 * Commitment: service index and hash pair for outputs.
 *
 * @param serviceIndex Service index
 * @param hash Output hash (32 bytes)
 */
final case class Commitment(
  serviceIndex: Long,
  hash: JamBytes
)

/**
 * Result of single-service accumulation as defined in Gray Paper equation 291.
 *
 * @param postState Modified state after accumulation
 * @param deferredTransfers Outgoing transfers queued during accumulation
 * @param yieldHash Accumulation output (32-byte hash or None)
 * @param gasUsed Actual gas consumed
 * @param provisions Service/blob pairs to provision
 */
final case class AccumulationOneResult(
  postState: PartialState,
  deferredTransfers: List[DeferredTransfer],
  yieldHash: Option[JamBytes],
  gasUsed: Long,
  provisions: Set[(Long, JamBytes)]
)

/**
 * Result of parallel accumulation.
 *
 * @param postState Combined modified state
 * @param deferredTransfers All outgoing transfers
 * @param outputs Set of (serviceIndex, hash) pairs
 * @param gasUsed Service to gas used mapping
 */
final case class AccumulationParResult(
  postState: PartialState,
  deferredTransfers: List[DeferredTransfer],
  outputs: Set[Commitment],
  gasUsed: List[(Long, Long)]
)

/**
 * Result of sequential accumulation.
 *
 * @param reportsAccumulated Number of reports accumulated
 * @param postState Final state after all accumulation
 * @param outputs Set of (serviceIndex, hash) pairs
 * @param gasUsed Service to gas used mapping
 */
final case class AccumulationSeqResult(
  reportsAccumulated: Int,
  postState: PartialState,
  outputs: Set[Commitment],
  gasUsed: List[(Long, Long)]
)
